File: TMP/Leetcode/sbgo/Leetcode583-两个字符串的删除操作.go
package sbgo

func minDistance(word1 string, word2 string) int {
	n, m := len(word1), len(word2)
	dp := make([][]int, n+1)
	for i := range dp {
		dp[i] = make([]int, m+1)
	}
	for i := 0; i <= n; i++ {
		dp[i][0] = i
	}
	for i := 0; i <= m; i++ {
		dp[0][i] = i
	}
	for i := 1; i <= n; i++ {
		for j := 1; j <= m; j++ {
			dp[i][j] = min(dp[i-1][j], dp[i][j-1]) + 1
			if word1[i-1] == word2[j-1] {
				dp[i][j] = min(dp[i-1][j-1], dp[i][j])
			}
		}
	}
	return dp[n][m]
}

func min(i, j int) int {
	if i < j {
		return i
	}
	return j
}

File: TMP/Leetcode/PY/Leetcode583-两个字符串的删除操作.py
class Solution:
    def minDistance(self, word1: str, word2: str) -> int:
        n,m = len(word1),len(word2)
        dp = [[0] * (m + 1) for _ in range(n + 1)]
        for i in range(n + 1): dp[i][0] = i
        for i in range(m + 1): dp[0][i] = i
        for i in range(1,n + 1):
            for j in range(1,m + 1):
                dp[i][j] = min(dp[i - 1][j],dp[i][j - 1]) + 1
                if word1[i - 1] == word2[j - 1]: dp[i][j] = min(dp[i][j],dp[i - 1][j - 1])
        return dp[n][m]